common/folderDeletionManager.go

```diff
@@ -0,0 +1,226 @@
+func (s *standardFolderDeletionManager) getParent(u *url.URL) (string, bool) {
+	if len(u.Path) == 0 {
+		return "", false // path is already empty, so we can't go up another level
+	}
+
+	// trim off last portion of path (or all of the path, if it only has one component)
+	c := s.clean(u)
+	lastSlash := strings.LastIndex(c, "/")
+	return c[0:lastSlash], true
+}
+
+// getStateAlreadyLocked assumes the lock is already held
+func (s *standardFolderDeletionManager) getStateAlreadyLocked(folder string) *folderDeletionState {
+	state, alreadyKnown := s.contents[folder]
+	if alreadyKnown {
+		return state
+	} else {
+		state = &folderDeletionState{}
+		s.contents[folder] = state
+		return state
+	}
+}
+
+func (s *standardFolderDeletionManager) RecordChildExists(childFileOrFolder *url.URL) {
+	folder, ok := s.getParent(childFileOrFolder)
+	if !ok {
+		return // this is not a child of any parent, so there is nothing for us to do
+	}
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	folderStatePtr := s.getStateAlreadyLocked(folder)
+	folderStatePtr.childCount++
+}
+
+func (s *standardFolderDeletionManager) RecordChildDeleted(childFile *url.URL) {
+	folder, ok := s.getParent(childFile)
+	if !ok {
+		return // this is not a child of any parent, so there is nothing for us to do
+	}
+
+	s.mu.Lock()
+	folderStatePtr, alreadyKnown := s.contents[folder]
+	if !alreadyKnown {
+		// we are not tracking this child, so there is nothing that we should do in response
+		// to its deletion (may happen in the recursive calls from tryDeletion, when they recurse up to parent dirs)
+		s.mu.Unlock()
+		return
+	}
+	folderStatePtr.childCount--
+	if folderStatePtr.childCount < 0 {
+		// should never happen. If it does it means someone called RequestDeletion and Recorded a child as deleted, without every registering the child as known
+		folderStatePtr.childCount = 0
+	}
+	deletionFunc := folderStatePtr.deleter
+	shouldDel := folderStatePtr.shouldDeleteNow()
+	s.mu.Unlock() // unlock before network calls for deletion
+
+	if shouldDel {
+		s.tryDeletion(folder, deletionFunc)
+	}
+}
+
+func (s *standardFolderDeletionManager) RequestDeletion(folder *url.URL, deletionFunc FolderDeletionFunc) {
+	folderStr := s.clean(folder)
+
+	s.mu.Lock()
+	folderStatePtr := s.getStateAlreadyLocked(folderStr)
+	folderStatePtr.deleter = deletionFunc
+	shouldDel := folderStatePtr.shouldDeleteNow() // test now in case there are no children
+	s.mu.Unlock()                                 // release lock before expensive deletion attempt
+
+	if shouldDel {
+		s.tryDeletion(folderStr, deletionFunc)
+	}
+}
+
+func (s *standardFolderDeletionManager) tryDeletion(folder string, deletionFunc FolderDeletionFunc) {
+	success := deletionFunc(s.ctx, s.logger) // for safety, deletionFunc should be coded to do nothing, and return false, if the directory is not empty
+
+	if success {
+		s.mu.Lock()
+		delete(s.contents, folder)
```

Comment:
Is this locking around a network call?

Reply:
Good question.  No, it's just locking around the delete from the in-memory map.